app/models/miq_filter.rb

```diff
@@ -101,11 +101,14 @@ def self.apply_belongsto_filters(inputs, bfilters)
     return []     if inputs.nil?
     return inputs if bfilters.empty?
 
+    vcmeta_index = bfilters.index_with { |tag| belongsto2object_list(tag) }
+
     filtered = []
     inputs.each do |p|
       bfilters.each do |tag|
+        vcmeta_list = vcmeta_index[tag]
+
         if p.kind_of?(Storage)
-          vcmeta_list = belongsto2object_list(tag)
           vcmeta_list.reverse_each do |vcmeta|
             if vcmeta.respond_to?(:storages) && vcmeta.storages.include?(p)
```

Comment:
Follow up PR, I think we can either a) preload this or even better b) preload storage_ids instead.

Reply:
would be cool if we didn't need this class lookup here. not sure if there is a way other code that relies upon his `kind_of?`/`Constantize``/find_by`.

maybe it is possible to know about these values without traversing it